### PR TITLE
Add MTBO tutorial

### PR DESCRIPTION
**Pull request by wjmaddox**
Summary: Add a MTBO tutorial demonstrating superiority of MTGPs over batch GPs on a composite problem. TODO: get outcome transforms working and MTGPs to perform better on this problem.

Differential Revision: D29599548

### CODE REVIEW

File: tutorials/composite_mtbo.ipynb [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "env: CUDA_LAUNCH_BLOCKING=1\n"
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:27]: let's get rid of this

File: tutorials/composite_mtbo.ipynb [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+    "\n",
+    "See [the composite BO tutorial w/ HOGP](https://github.com/pytorch/botorch/blob/master/tutorials/composite_bo_with_hogp.ipynb) for a more technical introduction. In general, we suggest using MTGPs for unstructured task outputs and the HOGP for matrix / tensor structured outputs.\n",
+    "\n",
+    "\n",
+    "We will use a Multi-Task Gaussian process ([MTGP](https://papers.nips.cc/paper/2007/hash/66368270ffd51418ec58bd793f2d9b1b-Abstract.html)) with an ICM kernel to model all of the outputs in this problem. MTGPs can be easily accessed in Botorch via the `botorch.models.KroneckerMultiTaskGP` model class (for fully observed outputs). Given $T$ tasks (outputs) and $n$ data points, they assume that the responses, $Y \\sim \\mathbb{R}^{n \\times T},$ are distributed as $\\text{vec}(Y) \\sim \\mathcal{N}(f, D)$ and $f \\sim \\mathcal{GP}(\\mu_{\\theta}, K_{XX} \\otimes K_{T}),$ where $D$ is a (diagonal) noise term."
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:15]:
```suggestion
    "We will use a Multi-Task Gaussian process ([MTGP](https://papers.nips.cc/paper/2007/hash/66368270ffd51418ec58bd793f2d9b1b-Abstract.html)) with an ICM kernel to model all of the outputs in this problem. MTGPs can be easily accessed in Botorch via the `botorch.models.KroneckerMultiTaskGP` model class (for the "block design" case of fully observed outputs at all inputs). Given $T$ tasks (outputs) and $n$ data points, they assume that the responses, $Y \\sim \\mathbb{R}^{n \\times T},$ are distributed as $\\text{vec}(Y) \\sim \\mathcal{N}(f, D)$ and $f \\sim \\mathcal{GP}(\\mu_{\\theta}, K_{XX} \\otimes K_{T}),$ where $D$ is a (diagonal) noise term."
```

File: tutorials/composite_mtbo.ipynb [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We use `GenericMCObjective` to define the differentiable function that we are optimizing. Here, it is defined as \n",
+    "$$g(f) = \\sum_{i=1}^T \\cos(f_i^2 + f_i w_i)$$\n",
+    "where $w$ is a weight vector (drawn randomly once at the start of the optimization). As this function is a non-linear function of the outputs $f,$ we cannot compute acquisition functions via computation of the posterior mean and variance, but rather have to compute posterior samples and evaluate acquisitions with Monte Carlo sampling. \n",
+    "\n",
+    "For greater than $10$ or so tasks, this would have been pretty much infeasible to sample the posterior over all tasks jointly, except that [Maddox et al, '21](https://arxiv.org/abs/2106.12997) have devised an efficient method for exploiting the structure in the posterior distribution of the MTGP, enabling efficient MC based optimization of objectives using MTGPs."
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:122]:
```suggestion
    "For greater than $10$ or so tasks, it is computationally challenging to sample the posterior over all tasks jointly using conventional approaches. However, [Maddox et al, '21](https://arxiv.org/abs/2106.12997) have devised an efficient method that exploits the structure in the posterior distribution of the MTGP, enabling efficient MC based optimization of objectives using MTGPs."
```

File: tutorials/composite_mtbo.ipynb [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "problem = ContextualHartmann6(noise_std = 0.001, negate=True).to(**tkwargs)\n",
+    "\n",
+    "weights = torch.randn(20, **tkwargs)\n",
+    "def callable_func(samples, X=None):\n",
+    "    res = -torch.cos((samples**2) + samples * weights)\n",
+    "    return res.sum(dim=-1)\n",
+    "objective = GenericMCObjective(callable_func)"
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:137]: nit: add some blank lines in between to make more readable

File: tutorials/composite_mtbo.ipynb [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def construct_acqf(model, objective, num_samples, best_f):\n",
+    "    sampler = IIDNormalSampler(num_samples=num_samples)\n",
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:187]: may want to comment on why you use IID sampling here

File: tutorials/composite_mtbo.ipynb [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+    "    return qEI"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_init = 20\n",
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:203]: suggest moving this into the SMOKE_TEST as well and choosing a smaller number to speed things up even more

File: tutorials/composite_mtbo.ipynb [hunk 7]
```diff
@@ -0,0 +1,740 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Automatic pdb calling has been turned ON\n"
+     ]
+    }
+   ],
+   "source": [
+    "%pdb"
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:233]: remove

File: tutorials/composite_mtbo.ipynb [hunk 8]
```diff
@@ -0,0 +1,740 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mtgp_results = torch.stack(mtgp_trial_objectives)[:, n_init:].cummax(1)[0]\n",
+    "batch_results = torch.stack(batch_trial_objectives)[:, n_init:].cummax(1)[0]\n",
+    "random_results = torch.stack(rand_trial_objectives)[:, n_init:].cummax(1)[0]"
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:633]: nit

```suggestion
    "mtgp_results = torch.stack(mtgp_trial_objectives)[:, n_init:].cummax(1).values\n",
    "batch_results = torch.stack(batch_trial_objectives)[:, n_init:].cummax(1).values\n",
    "random_results = torch.stack(rand_trial_objectives)[:, n_init:].cummax(1).values"
```

File: tutorials/composite_mtbo.ipynb [hunk 9]
```diff
@@ -0,0 +1,740 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Composite Bayesian Optimization with Multi-Task Gaussian Processes\n",
+    "\n",
+    "In this tutorial, we'll be describing how to perform multi-task Bayesian optimization over composite functions. In these types of problems, there are several related outputs, and an overall easy to evaluate objective function that we wish to maximize.\n",
+    "Multi-task Bayesian Optimization was first proposed by [Swersky et al, NeurIPS, '13](https://papers.neurips.cc/paper/2013/hash/f33ba15effa5c10e873bf3842afb46a6-Abstract.html) in the context of fast hyper-parameter tuning for neural network models; however, we demonstrate a more advanced use-case of [composite Bayesian optimization](https://proceedings.mlr.press/v97/astudillo19a.html) where the overall function that we wish to optimize is a cheap-to-evaluate (and known) function of the outputs. \n",
```
Comment (Balandat) [inline tutorials/composite_mtbo.ipynb:10]: Can you add a couple of sentences of motivation for why doing composite BO is valuable here? Using more information in the modeling yadda yadda

File: tutorials/composite_mtbo.ipynb (hunk 5 above)
Comment (wjmaddox) [reply tutorials/composite_mtbo.ipynb:187]: Actually I think we could use QMC here, will check.

Comment (wjmaddox) [reply tutorials/composite_mtbo.ipynb]: Wasn't seeing much / any improvement with QMC here so leaving as is.

**Pull request closed without merging**